File: Python/CreationalDesignPattern/Decorator_design.py
from abc import ABC, abstractmethod


# Product
class INotification(ABC):
    @abstractmethod
    def send(self, message: str):
        pass


class SMSNotification(INotification):
    def send(self, message: str):
        print(f"Sending SMS notification {message}")


class NotificationDecorator(INotification, ABC):
    def __init__(self, notification: INotification):
        self.base_notification = notification


class FormatNotification(NotificationDecorator):
    def __init__(self, notification: INotification):
        super().__init__(notification)

    def send(self, message: str):
        print("Notification is formatted")
        self.base_notification.send(message)


class AnimationNotification(NotificationDecorator):
    def __init__(self, notification: INotification):
        super().__init__(notification)

    def send(self, message: str):
        print("AnimationNotification is added")
        self.base_notification.send(message)


def main():
    notification = SMSNotification()
    notification.send("pradeeep")


if __name__ == "__main__":
    main()

File: C++/CreationalDesignPattern/Decorator_design.cpp
#include <iostream>
#include <memory>
#include <string>

// Product
struct INotification {
    virtual void send(const std::string& message) = 0;
    virtual ~INotification() = default;
};

struct SMSNotification : INotification {
    void send(const std::string& message) override {
        std::cout << "Sending SMS notification " << message << std::endl;
    }
};

class NotificationDecorator : public INotification {
protected:
    std::unique_ptr<INotification> baseNotification;

public:
    explicit NotificationDecorator(std::unique_ptr<INotification> notification)
        : baseNotification(std::move(notification)) {}
};

class FormatNotification : public NotificationDecorator {
public:
    explicit FormatNotification(std::unique_ptr<INotification> notification)
        : NotificationDecorator(std::move(notification)) {}

    void send(const std::string& message) override {
        std::cout << "Notification is formatted" << std::endl;
        baseNotification->send(message);
    }
};

class AnimationNotification : public NotificationDecorator {
public:
    explicit AnimationNotification(std::unique_ptr<INotification> notification)
        : NotificationDecorator(std::move(notification)) {}

    void send(const std::string& message) override {
        std::cout << "AnimationNotification is added" << std::endl;
        baseNotification->send(message);
    }
};

int main() {
    std::unique_ptr<INotification> notification = std::make_unique<SMSNotification>();
    notification->send("pradeeep");
    return 0;
}
